### PR TITLE
Adds Exception for inputs > maximum_source_length

### PR DESCRIPTION
**Pull request by Adamits**
- Adds warning for maximum_target_length.
- Adds log_warning function
- Addresses #40

### CODE REVIEW

File: yoyodyne/collators.py [hunk 1]
```diff
@@ -16,23 +20,70 @@ class Collator:
...
+        """Checks if a source length in the batch is greater than the
+        maximum allowed source length.
+
+        Args:
+            itemlist (List[datasets.Item]).
+
+        Raises:
+            LengthError.
+        """
+        batch_max_source_length = max([len(item.source) for item in itemlist])
```
Comment (kylebgorman) [inline yoyodyne/collators.py:63]: The square brackets here are otiose:

`... = max(len(item.source) for item in itemlist)`

File: yoyodyne/collators.py [hunk 2]
```diff
@@ -16,23 +20,70 @@ class Collator:
...
+
+    def _check_target_length(self, itemlist: List[datasets.Item]):
+        """Checks if a target length in the batch is greater than the
+        maximum allowed target length for inference. Logs a warning suggesting
+        that this probably indicates the max_target_length is too short.
+
+        Args:
+            itemlist (List[datasets.Item]).
+        """
+        batch_max_target_length = max([len(item.target) for item in itemlist])
```
Comment (kylebgorman) [inline yoyodyne/collators.py:78]: ditto

File: yoyodyne/train.py [hunk 3]
```diff
@@ -424,7 +433,12 @@ def main() -> None:
     train_set.index.write(index)
     util.log_info(f"Index: {index}")
     train_loader, dev_loader = get_loaders(
-        train_set, dev_set, args.arch, args.batch_size
+        train_set,
+        dev_set,
+        args.arch,
+        args.batch_size,
```
Comment (kylebgorman) [inline yoyodyne/train.py:439]: I think it may be time to write `_get_loaders_from_args` or whatever, and have that call `get_loaders`.

File: yoyodyne/collators.py [hunk 4]
```diff
@@ -16,23 +20,70 @@ class Collator:
...
+    max_source_length: int
+    max_target_length: int
 
-    def __init__(self, pad_idx, config: dataconfig.DataConfig, arch: str):
+    def __init__(
+        self,
+        pad_idx,
+        config: dataconfig.DataConfig,
+        arch: str,
+        max_source_length: int,
```
Comment (kylebgorman) [inline yoyodyne/collators.py:31]: Do you want to pass in sensible defaults here? They are both 128 at the flag level, I believe, which seems like a decent default.

File: yoyodyne/util.py [hunk 5]
```diff
@@ -16,6 +16,18 @@ def log_info(msg: str) -> None:
...
+def log_warning(msg: str) -> None:
+    """Logs a warning msg to sys.stderr.
+
+    We can additionally consider logging to a file, or getting a handle to
+    the PL logger.
+
+    Args:
+        msg (str): the message to log.
+    """
+    log_info(Warning(msg))
```
Comment (kylebgorman) [inline yoyodyne/util.py:28]: I am wondering if this is really useful to convert to a `Warning` object. I have not found them useful in general.

**Review by kylebgorman — APPROVED** [review]
Okay, I like this, it definitely takes care of the issue. 

Why do this in the collator rather than the dataset object? Just asking questions: I don't have any particular argument why it is better to do it in one or the other place.

File: tests/collator_test.py [hunk 6]
```diff
@@ -4,30 +4,42 @@
...
-        ("transformer", False, False, False),
+        "arch",
+        "has_features",
+        "has_target",
+        "expected_separate_features",
+        "max_src_length",
+        "max_tgt_length",
+    ],
+    [
+        ("feature_invariant_transformer", True, True, False, 128, 128),
```
Comment (kylebgorman) [inline tests/collator_test.py:16]: If you take my advice and make 128 the default for both, this won't be necessary, it occurs to me.

**Comment by Adamits** [comment]
> Okay, I like this, it definitely takes care of the issue.
> 
> Why do this in the collator rather than the dataset object? Just asking questions: I don't have any particular argument why it is better to do it in one or the other place.

Yeah I had a lot of trouble deciding. I think I had thought I could just check the batch tensor, rather than looping over List. But ultimately I do not do that, so maybe the `Dataset` would be best? I could add to the initialization in the `Item` class. Do you think that'd be better?

File: yoyodyne/collators.py (hunk 4 above)
Comment (Adamits) [reply yoyodyne/collators.py:31]: I never liked setting defaults in two different places. But since this could potentially be used as a CLI or a python library -- I have no idea what is typical here. I guess defaults here does make the most sense.

File: yoyodyne/util.py (hunk 5 above)
Comment (Adamits) [reply yoyodyne/util.py:28]: Do you suggest just doing 

```
util.log_info(Warning(msg))
```
in the collator then?

File: yoyodyne/collators.py (hunk 4 above)
Comment (kylebgorman) [reply yoyodyne/collators.py:31]: I don't love this either, but we could have `defaults.py` which could contain these values as constants. What do you think? (That would be a separate PR.)

(So far I have just repeated ourselves in most places since there are other reasons to have defaults, e.g., for testing, and I want this library to be very useable from Python too.)

File: yoyodyne/util.py (hunk 5 above)
Comment (kylebgorman) [reply yoyodyne/util.py:28]: Oh, no, I am suggesting that this should not raise a warning at all, it should just log it to stderr like other things.

**Comment by kylebgorman** [comment]
The way I was thinking about this, the checks should be on the padded data, since we're really checking whether we have _enough_ padding or not, with the understanding that there has to be some finite upper bound to this or the batches will grow too large. 

The collator is one place you could do this, since it sees the padding. The other place you could do this is [around here](https://github.com/CUNY-CL/yoyodyne/blob/c3154d1b913a64f4d445e5b17079a45bb03028de/yoyodyne/batches.py#L13), though you'd have some passing around to do to make that work. Do you think that'd be better?

**Comment by Adamits** [comment]
> The way I was thinking about this, the checks should be on the padded data, since we're really checking whether we have _enough_ padding or not, with the understanding that there has to be some finite upper bound to this or the batches will grow too large.
> 
> The collator is one place you could do this, since it sees the padding. The other place you could do this is [around here](https://github.com/CUNY-CL/yoyodyne/blob/c3154d1b913a64f4d445e5b17079a45bb03028de/yoyodyne/batches.py#L13), though you'd have some passing around to do to make that work. Do you think that'd be better?

Right, `PaddedTensor` was my first choice too, but I didn't like having to tell it whether it is a source or target tensor. I think I can figure out a clean way to do this though. I will rework this PR and you can see what you think.

**Comment by kylebgorman** [comment]
> Right, `PaddedTensor` was my first choice too, but I didn't like having to tell it whether it is a source or target tensor. I think I can figure out a clean way to do this though. I will rework this PR and you can see what you think.

If you want, keep this PR up so we can compare side-by-side. This seems like an important decision.

File: yoyodyne/collators.py (hunk 4 above)
Comment (Adamits) [reply yoyodyne/collators.py:31]: `defaults.py` sounds like a good solution.

**Comment by kylebgorman** [comment]
Closing in favor of #53.
